Public/Src/Cache/ContentStore/Distributed/NuCache/NuCacheCollectionUtilities.cs

```diff
@@ -268,14 +280,37 @@ public static IEnumerable<T> QueryAndOrderInPages<T>(this IEnumerable<T> origina
                         sourceHasItems = false;
                     }
                 }
-
-                if (queue.Count == 0)
+                else if (pool.Count == 0)
                 {
+                    Contract.Assert(!sourceHasItems);
                     yield break;
                 }
+                else
+                {
+                    Contract.Assert(pool.Count > 0);
 
-                yield return queue.Top;
-                queue.Pop();
+                    int minimumYieldSize;
+                    if (!sourceHasItems)
+                    {
+                        // If the enumerator has no elements left, we know that the queue has the best order possible,
+                        // so we yield everything.
+                        minimumYieldSize = pool.Count;
+                    }
+                    else
+                    {
+                        minimumYieldSize = (int)Math.Floor(removalFraction * pool.Count);
+                        // To guarantee termination, we always yield at least one element
+                        minimumYieldSize = Math.Max(minimumYieldSize, 1);
+                        // Never yield more than one page of results. This is about maintaining baseline accuracy for very large fractions.
+                        minimumYieldSize = Math.Min(minimumYieldSize, pageSize);
+                    }
+
+                    Contract.Assert(minimumYieldSize <= pool.Count);
+                    for (var i = 0; i < minimumYieldSize; ++i) {

```

Comment:
Given my disagreement (and/or wrongness) about what pattern of yield sizes is under this code, it would be great if you could run in debug and tell me what series of minimumYieldSizes you see.